source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+    Tuple, \
+    Type
+from fab.artifact import \
+    Artifact, \
+    FileType, \
+    State, \
+    New, \
+    Unknown, \
+    Analysed, \
+    Compiled, \
+    Linked
+from fab.tasks import Task
+from fab.database import \
+    SqliteStateDatabase, \
+    FileInfoDatabase
+
+
+class PathMap(object):
+    def __init__(self,
+                 pattern: str,
+                 filetype: Type[FileType],
+                 state: Type[State]):
+        self._pattern = pattern
+        self._filetype = filetype
+        self._state = state
+
+    @property
+    def pattern(self):
+        return self._pattern
+
+    @property
+    def filetype(self):
+        return self._filetype
+
+    @property
+    def state(self):
+        return self._state
+
+    def match(self, path: Path) -> bool:
+        matched = False
+        if re.match(self.pattern, str(path)):
+            matched = True
+        return matched
+
+
+class Engine(object):
+    def __init__(self,
+                 workspace: Path,
+                 target: str,
+                 pathmaps: List[PathMap],
+                 taskmap: Mapping[
+                     Tuple[Type[FileType], Type[State]],
+                     Task]) -> None:
+        self._workspace = workspace
+        self._target = target
+        self._pathmaps = pathmaps
+        self._taskmap = taskmap
+        self._database = SqliteStateDatabase(workspace)
+
+    @property
+    def target(self) -> str:
+        return self._target
+
+    def process(self,
+                artifact: Artifact,
+                discovery,
+                objects,
+                lock) -> List[Artifact]:
+
+        new_artifacts = []
+        new_discovery = {}
+        new_objects = []
+        # Identify tasks that are completely new
+        if (artifact.state is New
+                and artifact.filetype is Unknown):
+            # Use the pathmap list to work out the
+            # filetype and starting state
+            new_artifact = None
+            for pathmap in self._pathmaps:
+                if pathmap.match(artifact.location):
+                    new_artifact = Artifact(artifact.location,
+                                            pathmap.filetype,
+                                            pathmap.state)
+            # Assuming we found a match and were able
+            # to create the artifact, return it so that
+            # it can be added to the queue
+            if new_artifact is not None:
+                # Also store its hash in the file database
+                file_info = FileInfoDatabase(self._database)
+                file_info.add_file_info(artifact.location,
+                                        new_artifact.hash)
+                new_artifacts.append(new_artifact)
+
+        elif artifact.state is Analysed:
+
+            # Work out whether this artifact needs to be
+            # included in the build or not - tag it as having
+            # been seen so the system knows it is Analysed
```

Comment:
Surely if `artifact.sate is Analysed` then it has been analysed?

Reply:
Perhaps poor wording of the comment - `artifact.state` is only accessible to the process which is currently looking at it - whereas here I was meaning the `discovery` array which is what allows _other processes_ to know that the units this artifact contains have been analysed.  I'll try to rephrase (or am I misunderstanding?)